src/mobu/services/business/notebookrunner.py

```diff
@@ -149,46 +167,68 @@ def missing_services(self, notebook: Path) -> bool:
+            if not all_notebooks:
+                msg = "No notebooks found in {self._repo_dir}"
+                raise NotebookRepositoryError(msg, self.user.username)
+
+            filter_results = NotebookFilterResults(all=all_notebooks)
+            filter_results.excluded_by_dir = {
+                n for n in filter_results.all if self.is_excluded(n)
+            }
+            filter_results.excluded_by_service = {
+                n for n in filter_results.all if self.missing_services(n)
+            }
+
+            if self._notebooks_to_run:
+                requested = {
                     self._repo_dir / notebook
-                    for notebook in self.options.notebooks_to_run
-                ]
-                not_found = set(requested) - set(notebooks)
+                    for notebook in self._notebooks_to_run
+                }
+                not_found = requested - filter_results.all
                 if not_found:
                     msg = (
-                        f"These notebooks do not exist in {self._repo_dir}:"
-                        f" {not_found}"
+                        "Requested notebooks do not exist in"
+                        f" {self._repo_dir}: {not_found}"
                     )
                     raise NotebookRepositoryError(msg, self.user.username)
-                notebooks = requested
-                self.logger.debug(
-                    "Running with explicit list of notebooks",
-                    notebooks=notebooks,
+                filter_results.excluded_by_requested = (
+                    filter_results.all - requested
                 )
 
-            if not notebooks:
-                msg = "No notebooks found in {self._repo_dir}"
-                raise NotebookRepositoryError(msg, self.user.username)
-            random.shuffle(notebooks)
-        return notebooks
+            filter_results.runnable = (
+                filter_results.all
+                - filter_results.excluded_by_service
+                - filter_results.excluded_by_dir
+                - filter_results.excluded_by_requested
+            )
+            if bool(filter_results.runnable):
+                self.logger.info(
+                    "Found notebooks to run",
+                    filter_results=filter_results.model_dump(),
+                )
+            else:
+                self.logger.warning(
+                    "No notebooks to run after filtering!",
+                    filter_results=filter_results.model_dump(),
+                )
+
+        return filter_results
 
     def next_notebook(self) -> Path:
+        if not self._notebooks:
+            self._notebooks = self.find_notebooks()
         if not self._notebook_paths:
-            self._notebook_paths = self.find_notebooks()
+            self._notebook_paths = random.sample(
+                list(self._notebooks.runnable), k=len(self._notebooks.runnable)
+            )
```

Comment:
```suggestion
            self._notebook_paths = list(self._notebooks.runnable)
            random.shuffle(self._notebook_paths)
```
It was confusing to see sample used here with k equal to the length of the list. I think the above is what you intended?